controller/scheduler-v2/scheduler.go

```diff
@@ -0,0 +1,1190 @@
+		return
+	}
+
+	var err error
+	defer func() {
+		if err != nil {
+			log.Error("error handling job request", "err", err)
+		}
+		s.sendEvent(NewEvent(EventTypeJobRequest, err, req))
+	}()
+
+	log.Info("handling job request")
+	switch req.RequestType {
+	case JobRequestTypeUp:
+		// startJob sets the HostID on the request if successful
+		// otherwise it remains blank
+		err = s.startJob(req)
+	case JobRequestTypeDown:
+		err = s.stopJob(req)
+	default:
+		err = fmt.Errorf("unknown job request type: %s", req.RequestType)
+	}
+}
+
+func (s *Scheduler) RunPutJobs() {
+	log := fnLogger()
+	log.Info("starting job persistence loop")
+	strategy := attempt.Strategy{Delay: 100 * time.Millisecond, Total: time.Minute}
+	for {
+		job, ok := <-s.putJobs
+		if !ok {
+			log.Info("stopping job persistence loop")
+			return
+		}
+		jobLog := log.New("job.id", job.ID, "job.state", job.State)
+		jobLog.Info("persisting job")
+		err := strategy.RunWithValidator(func() error {
+			return s.PutJob(job)
+		}, httphelper.IsRetryableError)
+		if err != nil {
+			jobLog.Error("error persisting job", "err", err)
+		}
+	}
+}
+
+func (s *Scheduler) HandleLeaderChange(isLeader bool) {
+	log := fnLogger()
+	s.isLeader = isLeader
+	if isLeader {
+		log.Info("handling leader promotion")
+		s.triggerRectify()
+	} else {
+		log.Info("handling leader demotion")
+		// TODO: stop job restart timers
+	}
+	s.sendEvent(NewEvent(EventTypeLeaderChange, nil, isLeader))
+}
+
+func (s *Scheduler) sendDiffRequests(f *Formation, diff Processes) {
+	log := fnLogger("app.id", f.App.ID, "release.id", f.Release.ID)
+	for typ, n := range diff {
+		if n > 0 {
+			log.Info(fmt.Sprintf("requesting %d new job(s) of type %s", n, typ))
+			for i := 0; i < n; i++ {
+				req := NewJobRequest(f, JobRequestTypeUp, typ, "", random.UUID())
+				req.state = JobStateRequesting
+				s.jobs.AddJob(req.Job)
+				s.HandleJobRequest(req)
+			}
+		} else if n < 0 {
+			log.Info(fmt.Sprintf("requesting removal of %d job(s) of type %s", -n, typ))
+			for i := 0; i < -n; i++ {
+				req := NewJobRequest(f, JobRequestTypeDown, typ, "", "")
+				s.HandleJobRequest(req)
+			}
+		}
+	}
+}
+
+func (s *Scheduler) followHost(h utils.HostClient) {
+	if _, ok := s.hostStreams[h.ID()]; ok {
+		return
+	}
+
+	log := fnLogger("host.id", h.ID())
+	log.Info("streaming job events")
+	events := make(chan *host.Event)
+	stream, err := h.StreamEvents("all", events)
+	if err != nil {
+		log.Error("error streaming job events", "err", err)
+		return
+	}
+
+	log.Info("getting active jobs")
+	jobs, err := h.ListJobs()
+	if err != nil {
+		log.Error("error getting active jobs", "err", err)
+		return
+	}
+	log.Info(fmt.Sprintf("got %d active job(s) for host %s", len(jobs), h.ID()))
+
+	for _, job := range jobs {
+		s.handleActiveJob(&job)
+	}
+
+	s.hostStreams[h.ID()] = stream
+
+	s.triggerSyncFormations()
+
+	go func() {
+		for e := range events {
+			s.jobEvents <- e
+		}
+		// TODO: reconnect this stream unless unfollowHost was called
```

Comment:
We should either fix this, or add an anti-entropy method to make sure we are following all hosts (this will get disconnected when flynn-host is updated for example). A post merge issue is fine.